### ISSUE TITLE
Enhancement: Prompt Library

### ISSUE THREAD

**Issue by danny-avila** [danny-avila/LibreChat#3023]
### What features would you like to see added?

Prompt Library with advanced management system including:
- ability to create/use prompts easily
- ability for prompts to have "fill-in" variables
- ability for users to not only create but share their prompts globally
- quick and easy access of prompts for chats from side panel
- filtering of prompts
- advanced and optional versioning of prompts
- categorize/label prompts
- ability for admins to share prompts globally
- ability for admins to restrict whether or not users can create their own prompts
- ability for admins to add categories for prompts

### Features planned but may not be initially implemented

- ability to import/export prompts
- ability to use slash command for quick prompt access
- ability to set conversation starters
- ability for admins to set conversation starters for specific/all endpoints
- ability to attach presets to prompts and vice-versa
- ability to chain prompts
- ability to favorite/rate prompts
- eval for prompts
- prompt metadata (know which messages/conversations used them)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by OskarVismaDev** [comment]
This feature would be super powerful if these prompt presets in the library could have files attached! :)

**Comment by danny-avila** [comment]
> This feature would be super powerful if these prompt presets in the library could have files attached! :)

At the moment, this was not something planned for the initial release. I will definitely consider it, I've had some ideas on this already.

**Comment by TheRealAlexV** [comment]
Loving this feature so far. I can totally use this to replace superpowers gpt and move 100% to librechat now!

One thing I would love to see improved is the variable field. Would it be possible to paste in multi-line variables? For example I've used variables in the past to represent job descriptions which are generally 50+ lines. I can still paste in all the content into the current field, but I wonder if gpt or claude treat that data different from properly formatted data?

**Comment by danny-avila** [comment]
> I can still paste in all the content into the current field, but I wonder if gpt or claude treat that data different from properly formatted data?

it depends what you see in the resulting text above with the highlighted values. something like this (rough draft)?

![image](https://github.com/danny-avila/LibreChat/assets/110412045/7bedab80-1c2c-40c9-8efb-a34baa905333)

**Comment by danny-avila** [comment]
Prompt slash commands are now live on latest: https://github.com/danny-avila/LibreChat/pull/3219

**Comment by vicevolf** [comment]
I really appreciate your work on the prompts🥰, as I agree that it is something that needs attention. Although different models are fundamentally different, different prompts like those in the GPT store can greatly influence the behavior of AI, and even the quality of service. If high-quality prompts can be pre-made for users or shared among users, it should be able to reduce the cost of getting started and improve the user experience.

**Comment by avimar** [comment]
This is amazing! I wasn't sure what the "command" thing was but now I see it.

Are we able to get editable categories?

I see they are hardcoded into the endpoint and into the localization files and icons. I have no clue how you'd deal with the localization and icons.

Specifically, I'm missing "emotions" or "journaling" or "personal".

And maybe just the ability to _disable_ some?

**Comment by danny-avila** [comment]
> I see they are hardcoded into the endpoint and into the localization files and icons. I have no clue how you'd deal with the localization and icons.

Yes there is actually already some code to create your own categories. I will have to be judicious with what icons to have available, on top of the existing ones

**Comment by danny-avila** [comment]
I've closed this issue as the general feature is now implemented, and a release is being cut.

Features planned but may not be initially implemented. I encourage you all to try it and give feedback on improvements to be made features related to prompts. These are the most interesting to me, feel free to create separate issues for them to "vote" on it, as some have already been made.

- ability to import/export prompts
- ability to set conversation starters
- ability for admins to set conversation starters for specific/all endpoints
- ability to attach presets to prompts and vice-versa
- ability to chain prompts
- ability to favorite/rate prompts
